[REVIEW] Spectral clustering functionality in RAFT.

aschaffer opened the pull request:
This PR tracks work on https://github.com/rapidsai/cugraph/issues/868.

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

afender commented on CHANGELOG.md:
```diff
@@ -18,6 +19,7 @@
 ## New Features
 - Initial RAFT version
 - PR #3: defining raft::handle_t, device_buffer, host_buffer, allocator classes
+- PR #12: Spectral Clustering
```
This is already under 0.15, it should be removed from 0.14 log

aschaffer replied:
Looks like this already got changed as requested. Resolving, then.

afender replied:
Not sure the change went through, I still see line 22 as a duplicate of line 4

aschaffer replied:
Removed the duplicate.

afender commented on cpp/include/raft/sparse/cusparse_wrappers.h:
```diff
@@ -182,21 +182,337 @@ inline void cusparsecoosortByRow(  // NOLINT
+  const int* csrRowPtr, const int* csrColInd, const float* x, const float* beta,
+  float* y, cudaStream_t stream) {
+  CUSPARSE_CHECK(cusparseSetStream(handle, stream));
+  return cusparseScsrmv(handle, trans, m, n, nnz, alpha, descr, csrVal,
```
Looks like this one is deprecated in CUDA 11.0 (along with a few others). 
Are we addressing this in this PR or separately?

https://docs.nvidia.com/cuda/cusparse/index.html

aschaffer replied:
It's not clear what this comment refers to: `cusparseSetStream()` is being deprecated (it doesn't seem that's the case, based on the documentation linked above); or, `cusparse<X>csrmv()` ? (which is taken care of: see the `#if ` (line 216) and `#else` (line 249).

afender replied:
My bad, I did miss the`#if` earlier in the file

afender commented on cpp/include/raft/spectral/cluster_solvers.hpp:
```diff
@@ -0,0 +1,66 @@
+ */
+#pragma once
+
+#include <raft/spectral/kmeans.hpp>
```
Created https://github.com/rapidsai/raft/issues/28 for this.

cjnolet commented on cpp/include/raft/spectral/kmeans.hpp:
```diff
@@ -0,0 +1,898 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
```
It would be nice to see this in a more shared location, outside of spectral. Maybe `raft/cluster/kmeans.hpp`? We can keep `cluster_solver_t` in spectral, but I think it'll also make it easier to integrate cuml's kmeans once it's moved over.

aschaffer replied:
But what if both must coexist? Then we'd have a `cugraph`-specific `kmeans` and a `cuml` one. Which should be kept separately, no? 

In any case, I can make the move once it's been clarified where. Please specify the desired location to create and move to. `raft/cluster/kmeans.hpp` confirmed?

cjnolet replied:
I'm okay keeping this one where it is until we figure out what path we plan to take with kmeans.

cjnolet commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+ * limitations under the License.
+ */
+
+#pragma once
```
It would also be nice to see this placed in a more shared location / namspace. `raft/linalg/lanczos.hpp`? (Really just thinking of Scipy's [eigsh](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.eigsh.html))

cjnolet replied:
Speaking of which- it would be really useful to have a `lobpcg` primitive available at some point. cc @afender

aschaffer replied:
Please specify the desired location to create and move to. `raft/linalg/lanczos.hpp` confirmed?

cjnolet replied:
Sure, I think `raft/linalg/lanczos.hpp` is fine. We could always move it down the road if we need to.

aschaffer replied:
Done.

cjnolet commented on cpp/include/raft/spectral/spectral_util.hpp:
```diff
@@ -0,0 +1,235 @@
+  // find next power of 2
+  p2m = next_pow2<IndexType_>(m);
+  // setup launch configuration
+  nthreads.x = max(2, min(p2m, 32));
```
Why not use the `dim3` constructor on one-line?

aschaffer replied:
Because:

1. there are `narrowing conversion` warnings due to `dim3{}` constructor expecting the unsigned version of whatever index_type is; and doing an explicit cast to `unsigned index_type` would fail when / if `index_type` _is_ unisgned already;
2. `nthreads.y` is calculated in terms of `nthreads.x`, which would not work if these become constructor arguments;

aschaffer replied:
OK, fixed.

cjnolet commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+  return 0;
+}
+
+/// Find Householder transform for 3-dimensional system
```
I guess this is all done primarily on host because of the tiny data size?

aschaffer replied:
Yes.

cjnolet commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+namespace spectral {
+
+// curandGeneratorNormalX
+inline curandStatus_t curandGenerateNormalX(curandGenerator_t generator,
```
Just a note: we (cuml) will soon be bringing over a nice `rng` API built on top curand that should remove the need to invoke curand directly.

aschaffer commented on cpp/include/raft/graph.hpp:
```diff
@@ -0,0 +1,550 @@
+ * limitations under the License.
+ */
+#pragma once
+/// #include <comms_mpi.hpp> // TODO: clarify what must be done about `comm`
```
This particular `TODO` has been clarified (there should be no dependency on `comms_mpi.hpp` in raft). I'll just remove the comment.

aschaffer commented on cpp/include/raft/graph.hpp:
```diff
@@ -0,0 +1,550 @@
+   *
+   * @param[out] degree                Device array of size V (V is number of vertices) initialized
+   * to zeros. Will contain the computed degree of every vertex.
+   * @param[in]  x                     Integer value indicating type of degree calculation
```
See previous (resolved) comment: this header has been removed from raft (see discussion https://nvidia.slack.com/archives/CV0B740TW/p1593703758057400?thread_ts=1593701441.056900&cid=CV0B740TW)

aschaffer commented on cpp/include/raft/graph.hpp:
```diff
@@ -0,0 +1,550 @@
+   *             GraphCSRView does not own the memory used to represent this graph. This
+   *             function does not allocate memory.
+   *
+   * @param  offsets               This array of size V+1 (V is number of vertices) contains the
```
See previous (resolved) comment: this header has been removed from raft (see discussion https://nvidia.slack.com/archives/CV0B740TW/p1593703758057400?thread_ts=1593701441.056900&cid=CV0B740TW)

aschaffer commented on cpp/include/raft/graph.hpp:
```diff
@@ -0,0 +1,550 @@
+   *             GraphCOOView does not own the memory used to represent this graph. This
+   *             function does not allocate memory.
+   *
+   * @param  source_indices        This array of size E (number of edges) contains the index of the
```
See previous (resolved) comment: this header has been removed from raft (see discussion https://nvidia.slack.com/archives/CV0B740TW/p1593703758057400?thread_ts=1593701441.056900&cid=CV0B740TW)

aschaffer commented on cpp/include/raft/spectral/kmeans.hpp:
```diff
@@ -0,0 +1,898 @@
+// =========================================================
+
+/// Compute distances between observation vectors and centroids
+/** Block dimensions should be (warpSize, 1,
```
Fixed.

aschaffer commented on cpp/include/raft/spectral/kmeans.hpp:
```diff
@@ -0,0 +1,898 @@
+ *  @param clusterSizes (Output, k entries) Number of points in each
+ *    cluster. Entries must be initialized to zero.
+ */
+template <typename IndexType_>
```
Done.

aschaffer commented on cpp/include/raft/spectral/kmeans.hpp:
```diff
@@ -0,0 +1,898 @@
+  auto stream = handle.get_stream();
+
+  // Compute cumulative sum of distances
+  thrust::inclusive_scan(thrust_exec_policy, thrust::device_pointer_cast(dists),
```
No. I have no way of enforcing that constraint. We've discussed this. `thrust_exec_policy` is given by the caller, which cannot be prevented from passing any thrust execution policy they want.

aschaffer commented on cpp/include/raft/spectral/kmeans.hpp:
```diff
@@ -0,0 +1,898 @@
+  CUDA_TRY(cudaMemcpy(&distsSum, distsCumSum + n - 1, sizeof(ValueType_),
+                      cudaMemcpyDeviceToHost));
+
+  // Randomly choose observation vector
```
What isn't needed? `distsSum` is needed (used below). Hence, it needs to be updated from device.

aschaffer commented on cpp/include/raft/spectral/kmeans.hpp:
```diff
@@ -0,0 +1,898 @@
+                         thrust::device_pointer_cast(dists + n),
+                         thrust::device_pointer_cast(distsCumSum));
+  CHECK_CUDA(stream);
+  CUDA_TRY(cudaMemcpy(&distsSum, distsCumSum + n - 1, sizeof(ValueType_),
```
Done.

aschaffer commented on cpp/include/raft/spectral/kmeans.hpp:
```diff
@@ -0,0 +1,898 @@
+  //linear interpolation logic:
+  //{
+  ValueType_ minSum{0};
+  CUDA_TRY(cudaMemcpy(&minSum, distsCumSum, sizeof(ValueType_),
```
Done.

aschaffer commented on cpp/include/raft/spectral/kmeans.hpp:
```diff
@@ -0,0 +1,898 @@
+  CUDA_TRY(cudaMemsetAsync(codes, 0, n * sizeof(IndexType_), stream));
+
+  // Choose first centroid
+  thrust::fill(thrust_exec_policy, thrust::device_pointer_cast(dists),
```
Same answer as above: no way of enforcing that inside the called function.

aschaffer commented on cpp/include/raft/spectral/kmeans.hpp:
```diff
@@ -0,0 +1,898 @@
+  blockDim_warp.x = WARP_SIZE;
+  blockDim_warp.y = 1;
+  blockDim_warp.z = BSIZE_DIV_WSIZE;
+  gridDim_warp.x = min((d + WARP_SIZE - 1) / WARP_SIZE, 65535);
```
Don't know: I haven't authored these kernels. Can you please explain what is desired here?

aschaffer replied:
Also, I'm getting `narrowing conversion` warnings with some of the `dim3()` constructor changes.

aschaffer replied:
Addressed all such issues except this one that triggers warnings when constructor is used (and does not, otherwise).

cjnolet replied:
I see 65535 hardcoded in several places so I am just suggesting using USHRT_MAX instead.

aschaffer replied:
The problem is `USHRT_MAX` is not guaranteed to be `65535` on _all_ platforms: http://www.cplusplus.com/reference/climits/ . See the footnote reference:

"SHRT_MAX | Maximum value for an object of type unsigned short int | 65535 (216-1) or greater*
* the actual value depends on the particular system and library implementation, but shall reflect the limits of these types in the target platform."

aschaffer replied:
I've replaced `65535` by `constexpr index_type_t grid_lower_bound{65535};` which is compile-time evaluated (i.e., better than a runtime global variable). I hope this is acceptable. Thanks in advance.

aschaffer commented on cpp/include/raft/spectral/kmeans.hpp:
```diff
@@ -0,0 +1,898 @@
+
+  // Compute distance between centroids and observation vectors
+  CUDA_TRY(cudaMemsetAsync(dists, 0, n * k * sizeof(ValueType_), stream));
+  blockDim.x = WARP_SIZE;
```
I'm not sure what is meant here by "consolidate". Could you please elaborate?

cjnolet replied:
Rather than redefining the same dim3  everywhere for grid and block sizes, my suggestion is have a utility function or just a constant somewhere.

aschaffer replied:
Still not clear, I'm afraid. It's not the same `dim3`, the values are slightly different. But even if they where, are you suggesting defining a global `dim3` just to avoid repetition? Polluting the namespace with global variables like that is not good practice. Or, maybe I'm misunderstanding.

Could you please post an example of what is desired here? Thanks in advance.

aschaffer commented on cpp/include/raft/spectral/kmeans.hpp:
```diff
@@ -0,0 +1,898 @@
+                          thrust::raw_pointer_cast(obs_copy), n, stream));
+
+  // Cluster assigned to each observation matrix entry
+  thrust::sequence(thrust_exec_policy, rows, rows + d * n);
```
We cannot make that assumption. Again, that constraint can only be imposed by the caller. The caller should document that at call point.

cjnolet replied:
The main reason I'm saying we should at least document the requirement / implication here is because of the thrust exec policy and the stream on the handle are not the same then this algorithm could break in very unexpected ways. Parts like the above are not expecting to run concurrently, so for example if we have a cudaMemcpyAsync in one stream and a bunch of thrust operations in another then this could give unexpected behaviors. Best to document this the requirement here so the user knows this implication.

aschaffer replied:
Okay, done.

aschaffer commented on cpp/include/raft/spectral/kmeans.hpp:
```diff
@@ -0,0 +1,898 @@
+  ValueType_ dist_new_private;
+
+  // Each row is processed by a thread
+  i = threadIdx.x + blockIdx.x * blockDim.x;
```
Again, I haven't authored these kernels, so cannot answer this. I'm not sure what is the issue with this code. Can you please elaborate?

cjnolet replied:
For example, why define the variable above just to assign it a few lines after? i only takes on one value in this kernel, why not define and assign it on the same line?

aschaffer replied:
Okay. Done.

aschaffer commented on cpp/include/raft/spectral/kmeans.hpp:
```diff
@@ -0,0 +1,898 @@
+    thrust::reduce(thrust_exec_policy, thrust::device_pointer_cast(dists),
+                   thrust::device_pointer_cast(dists + n));
+
+  return 0;
```
The idea was that other cluster solvers may return non-zeros. Hence the non-void return. I don't have any other reason for that. If `void` is desired here, I will make that change.

aschaffer commented on cpp/include/raft/spectral/kmeans.hpp:
```diff
@@ -0,0 +1,898 @@
+  CHECK_CUDA(stream);
+
+  // Divide sums by cluster size to get centroid matrix
+  blockDim.x = WARP_SIZE;
```
Done everywhere, except one place that triggers `narrowing conversion` warnings.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+ *    Workspace. Not needed if full reorthogonalization is disabled.
+ *  @return Zero if successful. Otherwise non-zero.
+ */
+template <typename IndexType_, typename ValueType_>
```
Fixed.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+                            lanczosVecs_dev + IDX(0, *iter, n), 1, stream));
+  }
+
+  CUDA_TRY(cudaDeviceSynchronize());
```
Fixed.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+}
+
+/// Perform one step of Francis QR algorithm
+/** Equivalent to two steps of the classical QR algorithm on a
```
Fixed.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+ *  @param n Matrix dimension.
+ *  @param iter Current Lanczos iteration.
+ *  @param iter_new Lanczos iteration after restart.
+ *  @param shiftUpper Pointer to upper bound for unwanted
```
`shiftUpper` / `shiflLower` are host. Added that to the comments.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+  memcpy(work_host, beta_host, (iter - 1) * sizeof(ValueType_));
+  Lapack<ValueType_>::sterf(iter, ritzVals_host, work_host);
+
+  // Debug: Print largest eigenvalues
```
I'd rather keep debug-related comments, if you don't mind. When debug time comes, they turn out to be very useful.

cjnolet replied:
Fair point. I think we should revisit once we have figured out how we should do logging in a unified way across cugraph/cuml. What do you think? For example, it would be really nice to have this always log to debug (or when debug mode is on).

aschaffer replied:
OK. We need to be careful, because many times "DEBUG"-decorated code comes with a bunch of necessary dependencies (`#include <iostream>` , etc.) which must be `#ifdef`-ed out (otherwise they pollute the namespace and increase compilation time). Plus...`#ifdef`-ed out code is ugly (uglier then...just comments). 

But I agree with adding more structure and discipline in the (near) future, when dealing with such things.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+
+  // Useful constants
+  const ValueType_ one = 1;
+  const ValueType_ zero = 0;
```
I'd rather not pollute the (global) namespace with such low-significance constants. In general I'd stay away from global declarations, in C++. For reusability, I'd use factories and singletons. But for such low-level constants that'd be overkill. I can declare them `constexpr` and they will become compile-time expressions.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+  ValueType_ *__restrict__ beta_host, ValueType_ *__restrict__ lanczosVecs_dev,
+  ValueType_ *__restrict__ work_dev, ValueType_ *__restrict__ eigVals_dev,
+  ValueType_ *__restrict__ eigVecs_dev, unsigned long long seed) {
+  using namespace spectral;
```
`using namespace` inside function scope is accepted practice (done in `cudf` and `cugraph`).

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+
+  // Estimate number of Lanczos iterations
+  //   See bounds in Kuczynski and Wozniakowski (1992).
+  // const ValueType_ relError = 0.25;  // Relative error
```
Removed them.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+
+  Lapack<ValueType_>::sterf(*effIter, alpha_host, beta_host);
+  *shift = -alpha_host[*effIter - 1];
+  // std::cout <<  *shift <<std::endl;
```
Fixed.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+
+  // Obtain tridiagonal matrix with Lanczos
+  *effIter = 0;
+  // maxIter_curr = min(maxIter, restartIter);
```
Fixed.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+                      nEigVecs * sizeof(ValueType_), cudaMemcpyHostToDevice));
+  // for (int i = 0; i < nEigVecs; ++i)
+  //{
+  //  std::cout <<*(work_host+(2*(*effIter)+i))<< std::endl;
```
Removed.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+  // Initialize random number generator
+  curandCreateGenerator(&randGen, CURAND_RNG_PSEUDO_PHILOX4_32_10);
+
+  // FIXME: This is hard coded, which is good for unit testing...
```
The `FIXME` has actually been fixed (added `seed`). Removed the stale comment.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+  for (i = *effIter; i < nEigVecs; ++i) work_host[i + 2 * (*effIter)] = 0;
+
+  // Copy results to device memory
+  CUDA_TRY(cudaMemcpy(eigVals_dev, work_host + 2 * (*effIter),
```
Fixed.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+
+  IndexType_ top_eigenparis_idx_offset = *effIter - nEigVecs;
+
+  // Debug : print nEigVecs largest eigenvalues
```
As mentioned in above post, I'd rather keep the commented debug blocks around, as they are useful when / if debugging becomes necessary.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+  ValueType_ *alpha_host = alpha_host_v.data();
+  ValueType_ *beta_host = beta_host_v.data();
+
+  //TODO: replace and fix allocation via RAFT handle
```
Outdated. Removed.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+
+  // Clean up and exit
+  curandDestroyGenerator(randGen);
+  return 0;
```
As mentioned above, I'd rather keep the return for a more generic interface. Let me elaborate on these reasons:

1. returning something makes the API more general; perhaps the API currently does _not_ return something but it might in the future;
2. this functions may become part of a more general interface, with different implementations (versions); these specific implementations may not return anything, but other implementations (that adhere to the same interface) might;
3. typically, this functions may return some measures (number of clusters, number of iterations, etc.); they don't currently but they might in the future; 
4. the original implementation had these return values; I'd rather err on the cautious side and assume that the original author had good reasons to return something;
5. returning a value may become useful for debugging; the return is currently trivial, but during a debug session the return could be hijacked to propagate useful information to the caller without changing the signature of the function;

cjnolet replied:
I understand the benefits and reasoning for having functions return values and I see your points above. Perhaps this promps a larger discussion about error handling & return types in RAFT. For example, we're about to move a bunch of primitives over from cuml and most if not all of these have void return types.

aschaffer replied:
In `cudf` the philosophy was to always return something (typically wrapped inside a `std::unique_ptr`). With C++1x return-value optimizations (RVO) there is no penalty in returning locals (in fact there might be performance gains, because the compiler automatically triggers the optimizers when it sees constructs like that).

So, should I `void` the above, or leave as they are (for now)?

cjnolet replied:
I'm fine leaving it like this until we have some more formal / more consistent agreement across RAFT about how we should do this. Most of the cuML prims that are in the process of being moved over have void return types, but in the absence of a formal agreement, I'm not really sure that's better than just returning 0 everywhere.

aschaffer commented on cpp/include/raft/spectral/lapack.hpp:
```diff
@@ -0,0 +1,343 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
```
LAPACK functions are called in `lanczos.hpp`.  The LAPACK functions are collected as `static` methods under `class Lapack`. So they already are under the umbrella of a namespace (the class).

aschaffer commented on cpp/include/raft/spectral/lapack.hpp:
```diff
@@ -0,0 +1,343 @@
+  // computes the QR factorization of a general matrix
+  static void geqrf(int m, int n, T *a, int lda, T *tau, T *work, int *lwork);
+  // Generates the real orthogonal matrix Q of the QR factorization formed by geqrf.
+  //static void orgqr( int m, int n, int k, T* a, int lda, const T* tau, T* work, int* lwork );
```
Removed.

aschaffer commented on cpp/include/raft/spectral/matrix_wrappers.hpp:
```diff
@@ -0,0 +1,356 @@
+    // Apply diagonal matrix
+    //
+    dim3 gridDim, blockDim;
+    gridDim.x = std::min((n + BLOCK_SIZE - 1) / BLOCK_SIZE, 65535);
```
Done.

aschaffer commented on cpp/include/raft/spectral/modularity_maximization.hpp:
```diff
@@ -0,0 +1,194 @@
+  auto stream = handle.get_stream();
+
+  std::tuple<vertex_t, weight_t, vertex_t>
+    stats;  //{iters_eig_solver,residual_cluster,iters_cluster_solver} // # iters eigen solver, cluster solver residual, # iters cluster solver
```
Not really. It's explaining what the `stats` tuple is about. But I simplified that comment.

aschaffer commented on cpp/include/raft/spectral/modularity_maximization.hpp:
```diff
@@ -0,0 +1,194 @@
+#include <raft/spectral/eigen_solvers.hpp>
+#include <raft/spectral/spectral_util.hpp>
+
+//#define COLLECT_TIME_STATISTICS 1
```
Done.

aschaffer commented on cpp/include/raft/spectral/modularity_maximization.hpp:
```diff
@@ -0,0 +1,194 @@
+
+    // Record results
+    modularity += partModularity;
+    // std::cout<< "partModularity " <<partModularity<< std::endl;
```
Removed.

aschaffer commented on cpp/include/raft/spectral/modularity_maximization.hpp:
```diff
@@ -0,0 +1,194 @@
+    modularity += partModularity;
+    // std::cout<< "partModularity " <<partModularity<< std::endl;
+  }
+  // modularity = modularity/nClusters;
```
Fixed.

aschaffer commented on cpp/include/raft/spectral/sm_utils.hpp:
```diff
@@ -0,0 +1,435 @@
+#endif
+}
+
+static __inline__ __device__ double atomicFPAdd(double *addr, double val) {
```
I don't know enough about raft's general design directions. I placed this header in `spectral/` because I assumed limited use (spectral and spectral-related functionality in `cugraph`). It doesn't seem to me there's enough reason to reuse these outside of spectral clustering. In fact, I suspect many would be deprecated / replaced in the (near) future.

But, if desired, I can move it somewhere. `raft/utils` (directory?) doesn't currently exist. Please specify exactly where this should be moved, because when that happens it will trigger necessary updates in the two other spectral clustering-related PRs. So, if necessary, I'd rather do this only once, rather than "experiment" with different locations, etc.

aschaffer replied:
IOW, if `raft/utils/sm_utils.hpp` _is_ the desired location then I will create the `utils` directory and move it. But, I'd prefer reviewers to be _sure_ that that is desired course of action, here. Thanks in advance.

cjnolet replied:
I think `raft/utils/sm_utils.hpp` is a great location. I've had to define some of these in the past and it's nice to have them around when needed (also so there isn't duplication when they are needed).

aschaffer replied:
Done.

aschaffer commented on cpp/include/raft/spectral/spectral_util.hpp:
```diff
@@ -0,0 +1,235 @@
+  mm = (((m + blockDim.x - 1) / blockDim.x) *
+        blockDim.x);  // m in multiple of blockDim.x
+  alpha = 0.0;
+  // printf("[%d,%d,%d,%d] n=%d, li=%d, mn=%d \n",threadIdx.x,threadIdx.y,blockIdx.x,blockIdx.y, n,
```
Removed.

aschaffer commented on cpp/include/raft/spectral/warn_dbg.hpp:
```diff
@@ -0,0 +1,23 @@
+
+//nope:
+//
+#define WARNING(message)                                                  \
```
Similar argument: this seems specific to spectral clustering. I don't see a justification to make it more visible. But if current location isn't good then please specify where this should be moved. Thanks in advance.

cjnolet replied:
From earlier responses, I think the use of unified logging will remove this.

aschaffer replied:
OK. Postponing any action on this, for now. Once logging will be available, a clean-up PR would address these.

aschaffer commented on cpp/test/cluster_solvers.cu:
```diff
@@ -0,0 +1,104 @@
+  using value_type = double;
+
+  handle_t h;
+  ASSERT_EQ(0, h.get_num_internal_streams());
```
Copied over from other tests. Removed.

aschaffer commented on cpp/test/cluster_solvers.cu:
```diff
@@ -0,0 +1,104 @@
+
+  value_type modularity{0};
+  EXPECT_ANY_THROW(
+    spectral::analyzeModularity(h, t_exe_p, sm, k, clusters, modularity));
```
They act as a way to link and run code without actually performing any real test (which are deferred to `cugraph` and `cuml` sides). This is because `raft` is header only and linking outside dependencies like LAPACK, etc. would require linkage, hence more than just header only. These here are just "pulse" tests and ways to check correct compilation.

aschaffer replied:
`EXPECT_ANY_THROW` means the algorithm is expected to throw, but not caring about exception details (the algorithm _should_ throw because it's being passed `nullptr` pointers, which should be rejected by `RAFT_EXPECTS()`.

aschaffer commented on cpp/test/cluster_solvers.cu:
```diff
@@ -0,0 +1,104 @@
+  value_type* eigvals{nullptr};
+  value_type* eigvecs{nullptr};
+
+  unsigned long long seed{100110021003};
```
Again, actual testing is done from `cugraph` and `cuml` side. The testing data here, like `seed`, are just placeholders to invoke the algorithm and have a way to compile it.

aschaffer commented on cpp/test/eigen_solvers.cu:
```diff
@@ -0,0 +1,117 @@
+
+  lanczos_solver_t<index_type, value_type> eig_solver{cfg};
+
+  EXPECT_ANY_THROW(
```
Same comment as above: these are proof-of-life and compilation tests (which do not require linking external dependencies like LAPACK). Actual testing is done on the `cugraph` and `cuml` sides.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+  for (i = 0; i < restartSteps; i += 2)
+    if (francisQRIteration(iter, shifts_host[i], shifts_host[i + 1], alpha_host,
+                           beta_host, V_host, work_host))
+      WARNING("error in implicitly shifted QR algorithm");
```
As above, these are spectral-clustering specific, not worth exposing IMHO outside of spectral clustering. But, if really desired I can move them. Please specify where, as these will trigger cascading updates in the two other spectral-clustering related PRs. 

Two possibilities have already been expressed: moving `warn_dbg` header to a `utils` directory; or, moving these `WARNING`s to `error.hpp`. Once a decision is made I'll act on it. Thanks in advance.

cjnolet replied:
I believe this too will benefit from our unified logging API.

aschaffer replied:
OK. Postponing any action on this for now. Once logging will be available, a clean-up PR might address these.

aschaffer commented on cpp/include/raft/spectral/lanczos.hpp:
```diff
@@ -0,0 +1,1228 @@
+  // -------------------------------------------------------
+  // Check that LAPACK is enabled
+  // -------------------------------------------------------
+  // Lapack<ValueType_>::check_lapack_enabled();
```
Removed.

JohnZed commented:
Just checking @aschaffer - is this ready for re-review?

aschaffer commented:
> 
> 
> Just checking @aschaffer - is this ready for re-review?

Not sure what that means. I am addressing reviews as they come. I'm addressing the fresh batch today.

aschaffer commented:
rerun tests.
